File: blog-backend/src/main/java/com/blogproject/blogproject/service/WriterStatisticsService.java
package com.blogproject.blogproject.service;

import com.blogproject.blogproject.entities.User;
import com.blogproject.blogproject.enums.UserRole;
import com.blogproject.blogproject.repository.ArticleRepository;
import com.blogproject.blogproject.repository.FavoriteRepository;
import com.blogproject.blogproject.repository.LessonRepository;
import com.blogproject.blogproject.repository.QuizRepository;
import com.blogproject.blogproject.repository.SkillRepository;
import com.blogproject.blogproject.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class WriterStatisticsService {

    private final SkillRepository skillRepository;
    private final ArticleRepository articleRepository;
    private final LessonRepository lessonRepository;
    private final FavoriteRepository favoriteRepository;
    private final QuizRepository quizRepository;
    private final UserRepository userRepository;

    /**
     * PB-97
     */
    public long getSkillsCount() {
        return skillRepository.count();
    }

    /**
     * PB-98
     */
    public long getWrittenArticlesCount() {
        return articleRepository.count();
    }

    /**
     * PB-99
     */
    public long getLessonsCount() {
        return lessonRepository.count();
    }

    /**
     * PB-100
     */
    public long getFavoriteSkillsCount() {
        return favoriteRepository.count();
    }

    /**
     * PB-101
     */
    public long getReadersCount() {
        return userRepository.findAll()
                .stream()
                .filter(user -> user.getRole() == UserRole.READER)
                .count();
    }

    /**
     * PB-102
     */
    public long getQuizzesCount() {
        return quizRepository.count();
    }

    /**
     * Dashboard statistics
     */
    public Map<String, Long> getAllStatistics() {

        Map<String, Long> stats = new HashMap<>();

        stats.put("skills", getSkillsCount());
        stats.put("articles", getWrittenArticlesCount());
        stats.put("lessons", getLessonsCount());
        stats.put("favoriteSkills", getFavoriteSkillsCount());
        stats.put("readers", getReadersCount());
        stats.put("quizzes", getQuizzesCount());

        return stats;
    }
}

File: blog-frontend/src/app/writer-ui/services/writer-statistics.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface DashboardStatistics {
  skills: number;
  articles: number;
  lessons: number;
  favoriteSkills: number;
  readers: number;
  quizzes: number;
}

@Injectable({
  providedIn: 'root'
})
export class WriterStatisticsService {

  private http = inject(HttpClient);

  private readonly apiUrl =
    'http://localhost:8083/api/writer/statistics';

  getSkillsCount(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/skills`);
  }

  getArticlesCount(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/articles`);
  }

  getLessonsCount(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/lessons`);
  }

  getFavoriteSkillsCount(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/favorite-skills`);
  }

  getReadersCount(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/readers`);
  }

  getQuizzesCount(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/quizzes`);
  }

  getDashboardStatistics(): Observable<DashboardStatistics> {
    return this.http.get<DashboardStatistics>(
      `${this.apiUrl}/dashboard`
    );
  }
}
